### PR TITLE
individual error messages for invalid form inputs + added backend error message

### PR DESCRIPTION
**Pull request by amandahrdmn**

### CODE REVIEW

File: public/js/saveApplicant.js [hunk 1]
```diff
@@ -1,27 +1,62 @@
...
-    e.preventDefault()
-    let data = getCompletedFormData()
-    let validate = validateFormInputs(data)
-    if (validate) {
-        makeApiRequest(data)
+    e.preventDefault();
+    let data = getCompletedFormData();
+    let validate = validateFormInputs(data);
+    let formIsValid = true;
+    Object.keys(validate).forEach(formItem => {
```
Comment (bradleypurdue) [inline public/js/saveApplicant.js:6]: Can we get a line break before this please?

File: public/js/saveApplicant.js [hunk 2]
```diff
@@ -1,27 +1,62 @@
...
+        let itemKey = formItem;
+        let querySelector = `#${itemKey}Error`;
+        let formItemValues = validate[formItem];
+        Object.keys(formItemValues).forEach(validationType => {
```
Comment (bradleypurdue) [inline public/js/saveApplicant.js:10]: Can we get a line break before this please?

File: public/js/saveApplicant.js [hunk 3]
```diff
@@ -1,27 +1,62 @@
...
+let errorMessage = (validationType) => {
+    let htmlString = '';
+    if (validationType === 'isPresent') {
+        htmlString += `This is a required field, please fill in`
+    } else if (validationType === 'validLengthVarChar') {
+        htmlString += `This field must be less than 255 characters`
+    } else if (validationType === 'validLengthText') {
+        htmlString += `This field must be less than 10000 characters`
+    } else {
+        htmlString += `This field is invalid`
```
Comment (bradleypurdue) [inline public/js/saveApplicant.js:42]: Missing semi-colons.

File: public/js/hiringPartner.js [hunk 4]
```diff
@@ -3,7 +3,6 @@ const addContactForm = document.querySelector('#add-contact-form')
 
 addHiringPartnerForm.addEventListener('submit', async e => {
     e.preventDefault()
-    console.log('bing')
```
Comment (turnerror) [inline public/js/hiringPartner.js:6]: lol

File: public/js/saveApplicant.js [hunk 5]
```diff
@@ -32,26 +67,28 @@ let makeApiRequest = async (data) => {
...
-        })
+    }).then(response => {
+        if (response.status == 200) {
+            window.location.href = './admin';
+        } else if (response.status == 400) {
+            document.querySelector('#generalError').innerHTML = "You must fill out all form options.";
+            document.querySelector('#generalError').classList.remove('hidden');
+        } else {
+            document.querySelector('#generalError').innerHTML = "Something went wrong, please try again later.";
+            document.querySelector('#generalError').classList.remove('hidden');
```
Comment (turnerror) [inline public/js/saveApplicant.js:79]:
```suggestion
    }).then(response => {
     let ge = document.querySelector('#generalError');
        if (response.status == 200) {
            window.location.href = './admin';
        } else if (response.status == 400) {
            ge.innerHTML = "You must fill out all form options.";
            ge.classList.remove('hidden');
        } else {
            ge.innerHTML = "Something went wrong, please try again later.";
            ge.classList.remove('hidden');
```
more DRY. not perfect but smth like that

File: public/js/saveApplicant.js [hunk 6]
```diff
@@ -1,27 +1,62 @@
...
+            }
+        })
+    });
+
+    if (formIsValid) {
+        makeApiRequest(data);
     } else {
-        document.getElementById('errorMsg').innerHTML = 'Not all fields have been filled out. Try again...'
+        document.getElementById('generalError').innerHTML = 'This form is invalid, please check all fields';
+        document.querySelector('#generalError').classList.remove('hidden');
```
Comment (turnerror) [inline public/js/saveApplicant.js:30]: You are targetting the id generalError twice here yet using querySelector and getElementbyId. lets be consistent.

Comment (amandahrdmn) [reply public/js/saveApplicant.js:30]: yeah based off the old code and must have missed it - cheers!

File: public/js/saveApplicant.js [hunk 7]
```diff
@@ -32,26 +71,32 @@ let makeApiRequest = async (data) => {
...
+            window.location.href = './admin';
+        } else if (response.status === 400) {
+            generalErrorMessage.innerHTML = "You must fill out all form options.";
+            generalErrorMessage.classList.remove('hidden');
+            generalErrorMessage.classList.add('alert-danger');
+        } else {
+            generalErrorMessage = "Something went wrong, please try again later.";
+            generalErrorMessage.classList.remove('hidden');
+            generalErrorMessage.classList.add('alert-danger');
+        }
```
Comment (bradleypurdue) [inline public/js/saveApplicant.js:87]: You're removing and adding the same class on both of these. You could just do it once after the `else` statement is closed.

File: public/js/saveApplicant.js [hunk 8]
```diff
@@ -15,19 +21,16 @@ document.getElementById('submitApplicant').addEventListener('click', e => {
                 document.querySelector(querySelector).classList.remove('hidden');
                 document.querySelector(querySelector).innerHTML = errorMessage(validationType);
                 formIsValid = false;
-            } else {
-                document.querySelector(querySelector).classList.remove('alert-danger');
-                document.querySelector(querySelector).classList.add('hidden');
-                document.querySelector(querySelector).innerHTML = '';
```
Comment (turnerror) [inline public/js/saveApplicant.js]: this also could be a bit more DRY but tbh it reads well as well so Its up to you if you want to change it. I am happy either way.

File: public/js/saveApplicant.js (hunk 7 above)
Comment (amandahrdmn) [reply public/js/saveApplicant.js:87]: aha good shout

**Pull request merged**